src/http/request.rs

```diff
@@ -265,11 +270,24 @@ impl Request {
         Status::NGX_DONE
     }
 
+    /// how many subrequests are available to make in this request
+    /// will return NGX_HTTP_MAX_SUBREQUESTS for a parent request.
+    pub fn subrequests_available(&self) -> u32 {
+        // 1 is subtracted because this function was caught returning 1 extra
+        // The return value should be (50, 0), with the parent request returning
+        // NGX_HTTP_MAX_SUBREQUESTS.
+        // See http://nginx.org/en/docs/dev/development_guide.html#http_subrequests
+        // for more information
+        self.0.subrequests() - 1
+    }
+
     /// Send a subrequest
     pub fn subrequest(
         &self,
         uri: &str,
+        flags: u32,
```

Comment:
can we use enum instead?

#[derive(Debug)]
pub enum HttpSubrequestFlag {
    InMemory,
    Waited,
    Clone,
    Background,
}


then flags.into() would do the thing you would need to implement impl From<HttpSubrequestFlag> for ngx_uint_t { which will convert enum into nginx uint

Reply:
```rust
#[derive(Debug)]
pub enum NgxHttpSubrequestFlag {
    InMemory,
    Waited,
    Clone,
    Background,
}
// usize
impl From<NgxHttpSubrequestFlag> for ngx_uint_t {
    fn from(val: NgxHttpSubrequestFlag) -> Self {
        match val{
            NgxHttpSubrequestFlag::InMemory => NGX_HTTP_SUBREQUEST_IN_MEMORY as _,
            NgxHttpSubrequestFlag::Waited => NGX_HTTP_SUBREQUEST_WAITED as _,
            NgxHttpSubrequestFlag::Clone => NGX_HTTP_SUBREQUEST_CLONE as _,
            NgxHttpSubrequestFlag::Background => NGX_HTTP_SUBREQUEST_BACKGROUND as _,
        }
    }
}
```

then
```rust
pub fn subrequest(
        &self,
        uri: &str,
        module: &ngx_module_t,
        flag: NgxHttpSubrequestFlag,
        post_callback: unsafe extern "C" fn(*mut ngx_http_request_t, *mut c_void, ngx_int_t) -> ngx_int_t,
    )
// <    ... >
        let r = unsafe {
            ngx_http_subrequest(
                (self as *const Request as *mut Request).cast(),
                uri_ptr,
                std::ptr::null_mut(),
                &mut psr as *mut _,
                sub_ptr as *mut _,
                flag.into(),
            )
        };
// <    ... >
```

Reply:
We need to be able to support expressions like this for modules: `flags |= NGX_HTTP_SUBREQUEST_IN_MEMORY|NGX_HTTP_SUBREQUEST_WAITED;`

I think `bitflags` crate might be the most efficient and easy to use.

```
bitflags! {
  flags SubRequestFlags: u32 {
    const WAITED = 0x01,
    const InMemory = 0x02,
    const Clone = 0x04,
  }
}

let flag = WAITED | InMemory;
assert!(flag.contains(WAITED));
assert!(flag.contains(InMemory));
assert!(flag & Clone).is_empty());
```

Reply:
Although enum variants could work too:
```
enum SubRequestFlag {
  Waited = 0b0001,
  InMemory = 0b0010,
  Clone = 0b0100,
}
flags = Waited | InMemory
```

Reply:
If variants work how I hope, going to update accordingly.

Reply:
TIL: https://stackoverflow.com/questions/43509560/how-to-implement-bitwise-operations-on-a-bitflags-enum